### ISSUE TITLE
Feature Request: Per-Sample Connection-Level Regularization Layer for Keras

### ISSUE THREAD

**Issue by geleshChrsitUniversity** [keras-team/tf-keras#916]
Keras currently provides several activation-level regularization layers, including Dropout, GaussianDropout, AlphaDropout, and SpatialDropout. However, it does not provide a native layer for connection-level stochastic regularization.

_DropConnect (Wan et al., 2013)_ introduced connection-level masking as an alternative to activation-level dropout. Despite its established formulation, practical implementations have remained relatively uncommon because efficient implementations often require generating per-sample masked weight matrices or specialized backend implementations.

I would like to propose a Dense layer that performs per-sample connection-level stochastic masking using standard Keras tensor operations. The implementation does not require custom CUDA kernels or TensorFlow runtime changes and can be expressed entirely as a Keras layer.

A working proof-of-concept implementation, documentation, and experimental evaluation are available:

- PerNodeDrop repository: ([PerNodeDrop repository:](https://github.com/geleshChrsitUniversity/PerNodeDrop))
- Keras Discussion: ([idea](https://github.com/keras-team/keras/discussions/23090))
- Keras Discussion ([ShowAndTell] https://github.com/keras-team/keras/discussions/23237)
- Preprint of Pending publication (https://arxiv.org/abs/2512.12663)

The intention of this proposal is not to replace existing Dropout layers, but to provide native support for connection-level stochastic regularization within the Keras API. 

This feature would benefit users interested in exploring connection-level stochastic regularization without maintaining custom layers or backend-specific implementations.

It would provide a native Keras implementation of an established regularization concept, making experimentation easier for researchers, educators, and practitioners while remaining consistent with existing Keras workflows.

**Comment by geleshChrsitUniversity** [comment]
I think appropriate was Keras/issued
Hence Closing, porting it to https://github.com/keras-team/keras/issues/23246

**Comment by Abdellox** [comment]
Thanks for opening this! I'd like to help investigate.

Could you share:
- Steps to reproduce
- Expected vs actual behavior
- Environment details

I'll take a look at the code and see if I can help.

**Comment by geleshChrsitUniversity** [comment]
Alternative code

import tensorflow as tf
from tensorflow.keras import layers, initializers

_VALID_STIR_TYPES = ("drop", "gaussian")


class PerNodeDropDense(layers.Layer):
    """CORRECTED: true node-level (per-unit) masking.

    Mask shape is (batch, units) -- one perturbation per OUTPUT unit per
    sample -- applied to the activation AFTER a single, ordinary,
    shared-weight matmul. The weight matrix is never touched or
    duplicated per sample, so this preserves a standard batched GEMM and
    scales linearly (batch x units) rather than quadratically
    (batch x input_dim x units) in layer width.
    """

    def __init__(self, units, drop_rate=0.2, fixed_mask=False,
                 stir_type="drop", activation=None, **kwargs):
        super().__init__(**kwargs)
        self.units = units
        self.drop_rate = drop_rate
        self.fixed_mask = fixed_mask
        self.stir_type = stir_type.lower()
        if self.stir_type not in _VALID_STIR_TYPES:
            raise ValueError(f"stir_type must be one of {_VALID_STIR_TYPES}, got {stir_type!r}.")
        self.activation = tf.keras.activations.get(activation)

    def build(self, input_shape):
        last_dim = input_shape[-1]
        self.kernel = self.add_weight(name="kernel", shape=(last_dim, self.units),
                                       initializer=initializers.GlorotUniform(), trainable=True)
        self.bias = self.add_weight(name="bias", shape=(self.units,),
                                     initializer="zeros", trainable=True)

        if self.fixed_mask:
            def fixed_mask_initializer(shape, dtype=None):
                if self.stir_type == "drop":
                    keep = tf.cast(tf.random.uniform(shape) > self.drop_rate, dtype)
                    return keep / (1.0 - self.drop_rate)
                else:
                    stddev = tf.sqrt(self.drop_rate / (1.0 - self.drop_rate))
                    return tf.random.normal(shape, mean=1.0, stddev=stddev, dtype=dtype)

            self.fixed_mask_value = self.add_weight(
                name="fixed_mask_value", shape=(self.units,),
                initializer=fixed_mask_initializer, trainable=False)
        else:
            self.fixed_mask_value = None
        super().build(input_shape)

    def _sample_dynamic_mask(self, mask_shape, dtype):
        if self.stir_type == "drop":
            keep = tf.cast(tf.random.uniform(mask_shape) > self.drop_rate, dtype)
            return keep / (1.0 - self.drop_rate)
        stddev = tf.sqrt(self.drop_rate / (1.0 - self.drop_rate))
        return tf.random.normal(mask_shape, mean=1.0, stddev=stddev, dtype=dtype)

    def call(self, inputs, training=False, return_masks=False):
        rank = inputs.shape.rank
        if rank == 2:
            z = tf.matmul(inputs, self.kernel)
        elif rank == 3:
            z = tf.tensordot(inputs, self.kernel, axes=[[2], [0]])
        else:
            raise ValueError("Unsupported input rank.")
        z = z + self.bias
        output = self.activation(z) if self.activation is not None else z

        mask = None
        if training or self.fixed_mask:
            if self.fixed_mask and self.fixed_mask_value is not None:
                mask = self.fixed_mask_value
            else:
                batch = tf.shape(inputs)[0]
                if rank == 2:
                    mask_shape = tf.stack([batch, self.units])
                else:
                    time = tf.shape(inputs)[1]
                    mask_shape = tf.stack([batch, time, self.units])
                mask = self._sample_dynamic_mask(mask_shape, inputs.dtype)
            output = output * mask

        if return_masks:
            return output, mask
        return output

    def get_config(self):
        config = super().get_config()
        config.update({"units": self.units, "drop_rate": self.drop_rate,
                        "fixed_mask": self.fixed_mask, "stir_type": self.stir_type,
                        "activation": tf.keras.activations.serialize(self.activation)})
        return config
